syft/serde/serde.py

```diff
@@ -0,0 +1,476 @@
+functionality into three steps. When converting from a PySyft object (or collection
+of objects) to an object to be sent over the wire (a message), those three steps
+are (in order):
+
+1. Simplify - converts PyTorch objects to simple Python objects (using pickle)
+2. Serialize - converts Python objects to binary
+3. Compress - compresses the binary (Now we're ready send!)
+
+Inversely, when converting from a message sent over the wire back to a PySyft
+object, the three steps are (in order):
+
+1. Decompress - converts compressed binary back to decompressed binary
+2. Deserialize - converts from binary to basic python objects
+3. Detail - converts some basic python objects back to PyTorch objects (Tensors)
+
+Furthermore, note that there is different simplification/serialization logic
+for objects of different types. Thus, instead of using if/else logic, we have
+global dictionaries which contain functions and Python types as keys. For
+simplification logic, this dictionary is called "simplifiers". The keys
+are the types and values are the simplification logic. For example,
+simplifiers[tuple] will return the function which knows how to simplify the
+tuple type. The same is true for all other simplifier/detailer functions.
+
+By default, the simplification/detail operations expect Torch tensors. If the setup requires other
+serialization process, it can override the functions _serialize_tensor and _deserialize_tensor
+
+By default, we serialize using msgpack and compress using lz4.
+If different compressions are required, the worker can override the function _apply_compress_scheme
+"""
+import torch
+import msgpack
+import lz4
+from lz4 import (  # noqa: F401
+    frame,
+)  # needed as otherwise we will get: module 'lz4' has no attribute 'frame'
+import numpy
+import zstd
+
+import syft
+
+from syft.workers import AbstractWorker
+from syft.workers import VirtualWorker
+
+from syft.federated import Plan
+
+from syft.exceptions import CompressionNotFoundException
+from syft.exceptions import GetNotPermittedError
+from syft.exceptions import ResponseSignatureError
+
+
+from syft.frameworks.torch.tensors.decorators import LoggingTensor
+from syft.frameworks.torch.tensors.interpreters import AdditiveSharingTensor
+from syft.frameworks.torch.tensors.interpreters import MultiPointerTensor
+from syft.frameworks.torch import pointers
+
+from syft.serde.native_serde import (
+    _simplify_str,
+    _simplify_range,
+    _simplify_ellipsis,
+    _simplify_slice,
+    _detail_str,
+    _detail_range,
+    _detail_ellipsis,
+    _detail_slice,
+)
+
+from syft.serde.torch_serde import (
+    _detail_torch_tensor,
+    _detail_torch_parameter,
+    _detail_collection_tuple,
+    _detail_collection_list,
+    _detail_collection_set,
+    _detail_dictionary,
+    _detail_ndarray,
+    _detail_torch_device,
+    _detail_pointer_tensor,
+    _detail_log_tensor,
+    _detail_additive_shared_tensor,
+    _detail_multi_pointer_tensor,
+    _detail_plan,
```

Comment:
_simplify_<syft_object> or _detail_<syft_object> should be replaced by sy.<syft_object>.simplify or sy.<syft_object>.detail.

So we would use sy.Plan.simplify instead of _simplify_plan, sy.PointerTensor.simplify instead of _simplify_pointer_tensor

Reply:
One tiny question :
For me It is not possible to => create a method simplify within Plan class for example where the method simplify take as an object a Plan object type.

```
class Plan
      @staticmethod
      def simplify(plan: Plan) -> tuple:
           ....
```
Should i just use a basic object but we will lost the strong Type of Plan
```
class Plan
      @staticmethod
      def simplify(plan: object) -> tuple:
           ....
```
After that it might still have some circular import issue because of recursion call within methods simplify and detail

Reply:
What if it is not static?

Reply:
It looks like it is possible to do it with this notation:

```
class Plan
      @staticmethod
      def simplify(plan: "Plan") -> tuple:
           ....
```
I ll try this way

Reply:
Good!

Reply:
@mari-linhares 
> _simplify_<syft_object> or _detail_<syft_object> should be replaced by sy.<syft_object>.simplify or sy.<syft_object>.detail.
> 
> So we would use sy.Plan.simplify instead of _simplify_plan, sy.PointerTensor.simplify instead of _simplify_pointer_tensor

I believe I can achieve your requirement.
For that i need to put all functions from protected to public (mainly for _detail and simplify to detail and simplify).
Is it an issue?

Reply:
I don't think so :)

Reply:
Should I move the following _detail and _simplify : [torch_device, exception, worker, full_detail_worker, script_module] to the AbstractWorker class ?

Reply:
Should I move more  _detail and _simplify methods? 
Exceptions, workers, torch.device, torch.jit ?

Reply:
My opinion: exceptions and workers should live in their own files (like you did for plans); torch.device, torch.jit should live in torch_serde.py

Reply:
I do agree :)